Characterize user-facing log messages

blairconrad opened the pull request:
Contributes to #172, implementing

1. Capture INFO and worse log output in the lib.rs tests, to characterize the state of things now
2. Add additional tests at this level, to cover the various situations that can cause a change to not be absorbed

blairconrad commented on src/lib.rs:
```diff
@@ -826,37 +1315,50 @@ mod tests {
         let is_something_in_index = !nothing_left_in_index(&ctx.repo).unwrap();
         assert!(is_something_in_index);
+
+        log_utils::assert_log_messages_match(
```
Probably overkill. When I wrote it, I thought I'd be using it more. If we restrict ourselves to just checking the level and msg, this could go away.

tummychow replied:
yeah i think this is too fussy for my tastes. let's not make future test writers jump through this hoop - the message alone is good enough

blairconrad commented on src/lib.rs:
```diff
@@ -1060,6 +1469,35 @@ mod tests {
+        log_utils::assert_log_messages_are(
+            capturing_logger.visible_logs(),
+            vec![
+                &json!({"level": "WARN", "msg": "Could not find a commit to fix up, use --base to increase the search range."}),
```
whoops. Some of the lines exceed the 100 char limit that the repo seems to aim for. Later commits for my grand plan end up fixing this (often the messages are different anyhow). Can reformat if you like.

blairconrad replied:
Since no comment from you yet, pushing shortened log lines. I hope it doesn't cause confusion.

blairconrad commented:
I see some log-affecting PRs coming in. I don't think they'll conflict with this, but if they get merged, I'll assess and rebase.

tummychow reviewed (approved):
this works for me. i'd like to see a refactor to use `SlogValue`, but it doesn't have to be in this pr, maybe you even have a draft of that coming up. unfortunately it looks like the clippy lint fixes conflicted with this one, so it'll require a small rebase

blairconrad commented:
Thanks, @tummychow.

* rebased to resolve clippy changes
* removed `assert_log_messages_match`
* added log-checking for the two new tests that appeared around the "message" option

> i'd like to see a refactor to use SlogValue

Was not aware of SlogValue until your comment.
Will investigate.

Meanwhile, I'm pushing in case you want to merge (but I don't mind being stuck behind #177; whatever's easier for you).

blairconrad commented:
haha, oops. I'd pushed without the log-checking in the extra test methods, then got called away. Came back, added them, and was about to push the updates. But I'm too late. Sorry.

blairconrad commented:
> i'd like to see a refactor to use SlogValue

I've investigated and am not sure what you're looking for.

What are you hoping will use SlogValue? 
Do you want the messages to become formalized and use a LogDetails and StatisticsLogger?

I experimented with a future message like

```rust
#[derive(Clone, ExtLoggable, Serialize)]
#[LogDetails(
    Id = "101",
    Text = "No changes staged, even after auto-staging. Try adding something to the index.",
    Level = "WARN"
)]
struct NoChangesStaged;

...

    if nothing_left_in_index(repo)? {
        xlog!(ui_logger, NoChangesStaged);
        return Ok(());
    }

```

And it may be something that will work, but I'm coming up with a panic when it runs.

tummychow commented:
i don't really care about the statistics, but having typed logging objects would make the user-facing logs more consistent and testable. eg right now you have to write out the log message once at the callsite where it's logged, and again at the place where you're testing it. it would be better to get those things into a single place. if you can't get it working, that's fine, but i think it's worth an attempt

blairconrad commented:
I'll see what I can do.
